### PR TITLE
Adding a page on FFI

### PR DESCRIPTION
**Pull request by sfshaza2**
Staged: https://sz-flutter.firebaseapp.com/docs/development/platform-integration/c-interop

Fixes https://github.com/flutter/website/issues/2941

I was asked to migrate a page from the Flutter wiki to our site. This is not a topic I know much about, so please help me clarify any confusion, errors, or omissions. Once this page is live, we can forward the old page.

Another question: Will this also work with web?

Thanks!

cc @kwalrath @mit-mit

### CODE REVIEW

**Comment by sjindel-google** [comment]
Thanks for doing this! It does not work on the Web yet.

I forwarded you some feedback from users who are actually very confused by the documentation, and I think we should consider improving it before publishing it on the website.

/cc @mkustermann @dcharkes

**Comment by sfshaza2** [comment]
@sjindel-google, I would agree with the assessment that the wiki page is highly confusing. It appears that there are many variables when connecting to different kinds of libraries, and on different target platforms. Also, you can connect directly or by creating a plugin.

I would suggest that we narrow the options down and select a few recommended ways to do this - typical scenarios. With your input/feedback/guidance, we can restructure the page completely, and make it more focused on the steps required. But I'm not quite sure yet, where to begin.

**Comment by sjindel-google** [comment]
I tried to organize it by use cases originally, but I didn't really describe what those are. Here's what I had in mind:

Use case 0:

Binding against a "system" library which is shipped with the target device.
Examples include binding against SQLite (on iOS) or GLES (on Android).

Use case 1:

First-party library. This is a library which the App developer has and wants to include in the App. I think we should not address this use case directly since it's a special case of (2) and (3).

Use case 2:

Open-source third-party library. The developer wants to create a Flutter plugin which includes both C/C++/Obj-C and Dart source code.

Use case 3:

Closed-source third-party library. The developer wants to create a Flutter plugin which includes Dart source code (necessary) but distribute the C/C++ library in binary form.

We don't want to encourage this use case in general, and we definitely don't want users putting binary artifacts on Pub. However, we should still provide guidance on doing this with the binary artifacts published through a third-party website, because some important users (e.g. Realm) may need it.

**Comment by sfshaza2** [comment]
Hi, @sjindel-google! I've taken a stab at making the doc more clear. Can you take a look? I've embedded some questions for you in bold text.  Thanks!!

File: src/docs/development/platform-integration/c-interop.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
+---
+title: "Binding to native code using dart:ffi"
+description: "To use C code in your Flutter program, use the dart:ffi library (currently in preview)."
+---
+
+Flutter mobile can use the [dart:ffi][] library
+to call native C APIs. The _ffi_ acronym stands for
```
Comment (sjindel-google) [inline src/docs/development/platform-integration/c-interop.md]: _FFI_

File: src/docs/development/platform-integration/c-interop.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  $ flutter upgrade
+  ```
+  For more information on Flutter's channels,
+  see [Upgrading Flutter][].
+{{ site.alert.end }}
+
+API documentation is available from the Dart dev channel:
+[Dart API reference documentation][].
+
+## iOS and macOS
```
Comment (sjindel-google) [inline src/docs/development/platform-integration/c-interop.md]: I think a forward is needed, like:

```
Before your library or program can use the FFI to bind to native code, you must ensure that the native code is loaded and its symbols are visible to Dart.

Thus this page focuses on compiling, packaging and loading native code within a Flutter plugin or app.
```

File: src/docs/development/platform-integration/c-interop.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+  ```terminal
+  $ flutter channel dev
+  $ flutter upgrade
+  ```
+  For more information on Flutter's channels,
+  see [Upgrading Flutter][].
+{{ site.alert.end }}
+
+API documentation is available from the Dart dev channel:
+[Dart API reference documentation][].
```
Comment (sjindel-google) [inline src/docs/development/platform-integration/c-interop.md]: I think we need a brief introduction to linking at be start:

A native library can be linked into an app either dynamically or statically. A statically linked library is embedded into the app's executable image, and will be loaded when the app starts.

Symbols from a statically linked library can be loaded via `DynamicLibrary.executable` or `DynamicLibrary.process`.

A dynamically linked library, by contrast, is distributed in a separate file or folder within the app, and loaded on-demand. On Android, a dynamically linked library is distributed as a set of `.so` (ELF) files, one for each architecture. On iOS, it's distributed as a `.framework` folder.

A dynamically linked library can be loaded into Dart via `DynamicLibrary.open`.

File: src/docs/development/platform-integration/c-interop.md (hunk 1 above)
Comment (kwalrath) [reply src/docs/development/platform-integration/c-interop.md]: Technically it's an initialism, not an acronym. :) But I'd just say "_FFI_ stands for..."

**Comment by sjindel-google** [comment]
I've re-worked the wiki page into a "Walkthrough": https://github.com/flutter/flutter/wiki/Binding-to-native-code-via-FFI

Would you mind taking a look at that instead? I think it will be much more helpful to most users.

**Comment by sfshaza2** [comment]
@sjindel-google, please take another look! It's staged on:

Staged: https://sz-flutter.firebaseapp.com/docs/development/platform-integration/c-interop

File: src/docs/development/platform-integration/c-interop.md [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+systems about the native code so the code can be compiled
+and linked appropriately into the final application.
+
+You add the sources to the `ios` folder,
+because CocoaPods doesn't allow including sources
+above the podspec file, but Gradle allows you to point
+to the `ios` folder. It's not required to use the same
+sources for both iOS and Android;
+you may, of course, add Android-specific sources
+to the `android` folder and modify the `CMakeLists.txt`
```
Comment (sjindel-google) [inline src/docs/development/platform-integration/c-interop.md]: "modify `CMakeLists.txt` appropriately"

File: src/docs/development/platform-integration/c-interop.md [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+file appropriately.
+
+The FFI library can only bind against C symbols,
+so in C++ these symbols must be marked "extern C".
```
Comment (sjindel-google) [inline src/docs/development/platform-integration/c-interop.md]: `extern C`

File: src/docs/development/platform-integration/c-interop.md [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+You should also add attributes to indicate that the
+symbols are referenced from Dart,
+to prevent a linker from discarding the symbols
+during link-time optimization.
+
+For example:
+
+```
```
Comment (sjindel-google) [inline src/docs/development/platform-integration/c-interop.md]: This quote is Bash, a command-line which can be run from the root of the directly. Can we make it show up with bash highlighting in the website?

**Pull request merged**